org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReferenceExpression.java

```diff
@@ -1300,9 +1300,13 @@ public boolean sIsMoreSpecific(TypeBinding s, TypeBinding t, Scope scope) {
 		TypeBinding[] sParams = sSam.parameters;
 		TypeBinding[] tParams = tSam.parameters;
 		// Both must have the same number of parameters if we got this far
-		for (int i = 0; i < sParams.length; i++) {
-			if (TypeBinding.notEquals(sParams[i], tParams[i]))
-				return false;
+		if(sParams.length == tParams.length) {
```

Comment:
When Sasikanth wrote this loop, he was convinced "Both must have the same number of parameters if we got this far". Unfortunately he is no longer on the team, so we can't ask him why he thought so.

At least we should have an example that tells us, how Sasi's assumption can be violated.

A brief look at the spec lets me think if perhaps one of the methods is varargs, in which case returning `false` might be wrong, as the spec has specific treatment for varargs methods.

Additionally, at first glance the corresponding method in `LambdaExpression` might have the same issue.


@srikanth-sankaran  fyi.

Reply:
JLS §18.5.4. refers to §15.12.2.5 for this part.

In JLS version 8 (from which this method was first implemented) I found: 
> (where U1 ... Uk and R1 are the parameter types and return type of the function type of the capture of S, and V1 ... Vk and R2 are the parameter types and return type of the function type of T)

To me this seems to say that also JLS 8 assumed both function types to have the same number of parameters (k).

Interestingly, this parenthesis can no longer be found in JLS 20. But still in version 20, the definitions of ` P1, ..., Pn` and `P1', ..., Pn'` use the same `n`.

Looking at the part specific to method reference expressions, JLS 8 defined `for all i (1 ≤ i ≤ k), Ui is the same as Vi,`. Very interestingly, I cannot find anything like that in JLS 20. If that was an intentional deletion in JLS, then also we should deleted the code comparing parameter types.

For clarity, I'm referring to this:
[JLS 8:](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12.2.5)
> If e is an exact method reference expression ([§15.13.1](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.13.1)), then i) for all i (1 ≤ i ≤ k), Ui is the same as Vi, and ii) one of the following is true: ...

[JLS 20:](https://docs.oracle.com/javase/specs/jls/se20/html/jls-15.html#jls-15.12.2.5)
> e is an exact method reference expression ([§15.13.1](https://docs.oracle.com/javase/specs/jls/se20/html/jls-15.html#jls-15.13.1)), and one of the following is true: ...

Reply:
Is removing the code right solution? As it sill works for other versions like 1.8 and I checked the tests are failing.

Reply:
> Is removing the code right solution? As it sill works for other versions like 1.8 and I checked the tests are failing.

Look at the uses of `org.eclipse.jdt.internal.compiler.impl.CompilerOptions.complianceLevel` in the compiler. By checking the compliance levels, it is possible to conditionally alter the behavior. Eventually, the behavior should match corresponding javac versions (ideally unless it is accepted to a defective unhelpful behavior) and should come along with a regression test that establishes appropriate behavior at different JDK levels